### DISCUSSION TITLE
Pass user-defined parameter to MACRO

### DISCUSSION THREAD

**Discussion by Aprillmylmy** [iiasa/message_ix#611] [Q&A] [answered]
Dear developer, 

I am wondering if it is possible to pass a user-specific parameter (e.g., maxloop=500 or similar constant parameters) into macro explicitly from the Jupyter notebook, instead of making changes from GAMS?

Thanks very much,
Emily

**Comment by khaeru** [comment, marked as answer]
Hi Emily—thanks for the question!

Yes, this is supported. Please have a look at the following:
- [`Scenario.init_scalar()` or perhaps `.init_par()`](https://docs.messageix.org/projects/ixmp/en/latest/api.html#ixmp.Scenario)—these methods allow to define a new 0-d or 1+-dimensional parameter, like your "maxloop", so that data can then be stored with `.change_scalar()` or `.add_par()`, respectively.
- When `scenario.solve()` is called, all data (sets and parameters) are written to GDX file. This will include your newly-defined one.
- You will still need to modify the GAMS code to retrieve that value from the GDX file in order to be available as a GAMS variable for use by other code. You should do that around here: https://github.com/iiasa/message_ix/blob/b572ccee45c293a88f1f2acf91a0811d50cf6593/message_ix/model/MACRO/macro_data_load.gms#L122-L139
  After having done this once, then afterwards you can use the Python API to change the value and every `.solve()` call will pass your custom parameter value through to the  GAMS code.

I hope that helps! Please mark the answer ✅ if it covered your initial question, and either add a reply or start a new thread if you have others.
